File: app/src/main/java/net/ballmerlabs/scatterbrain/ui/power/PowerFragment.java
    package net.ballmerlabs.scatterbrain.ui.power;

    import android.content.ComponentName;
    import android.content.Intent;
    import android.content.ServiceConnection;
    import android.os.Bundle;
    import android.os.IBinder;
    import android.os.RemoteException;
    import android.util.Log;
    import android.view.LayoutInflater;
    import android.view.View;
    import android.view.ViewGroup;
    import android.widget.TextView;
    import android.widget.ToggleButton;

    import androidx.annotation.NonNull;
    import androidx.fragment.app.Fragment;

    import net.ballmerlabs.scatterbrain.R;
    import net.ballmerlabs.scatterbrainsdk.ScatterbrainAPI;
    import net.ballmerlabs.uscatterbrain.ScatterRoutingService;

public class PowerFragment extends Fragment {
    private static final String TAG = "PowerFragment";
    private ScatterbrainAPI binder;
    private ToggleButton enableDisableRouterButton;
    private ServiceConnection mServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            binder = ScatterbrainAPI.Stub.asInterface(service);
            Log.v(TAG, "connected to ScatterRoutingService binder");
            try {
                if (!binder.isDiscovering())
                    binder.startDiscovery();
                updateCheckedStatus();
            } catch (RemoteException e) {
                Log.e(TAG, "RemoteException: " + e);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            updateCheckedStatus();
            binder = null;
        }
    };

    private synchronized void updateCheckedStatus() {
        if (enableDisableRouterButton == null) {
            return;
        }

        try {
            if (binder != null) {
                enableDisableRouterButton.setChecked(binder.isDiscovering() || binder.isPassive());
            }
        } catch (RemoteException e) {
            Log.e(TAG, "RemoteException: " + e);
            enableDisableRouterButton.setChecked(false);
        }
    }

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_power, container, false);
        final TextView textView = root.findViewById(R.id.text_home);

        enableDisableRouterButton = root.findViewById(R.id.toggleButton);

        final Intent bindIntent = new Intent(getActivity().getApplicationContext(), ScatterRoutingService.class);
        getActivity().bindService(bindIntent, mServiceConnection, 0);
        updateCheckedStatus();
        enableDisableRouterButton.setOnCheckedChangeListener((compoundButton, b) -> {
            final Intent startIntent = new Intent(getActivity().getApplicationContext(), ScatterRoutingService.class);
            if (b) {
                getActivity().startForegroundService(startIntent);
            } else {
                getActivity().unbindService(mServiceConnection);
                getActivity().stopService(startIntent);
            }
        });
        return root;
    }
}

File: app/src/main/java/net/ballmerlabs/scatterbrain/ui/power/PowerFragment.kt
package net.ballmerlabs.scatterbrain.ui.power

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.isActive
import net.ballmerlabs.scatterbrain.RoutingServiceViewModel
import net.ballmerlabs.scatterbrain.ServiceConnectionRepository
import net.ballmerlabs.scatterbrain.databinding.FragmentPowerBinding
import net.ballmerlabs.scatterbrain.softCancelLaunch
import javax.inject.Inject

@AndroidEntryPoint
class PowerFragment : Fragment() {

    @Inject lateinit var serviceConnectionRepository: ServiceConnectionRepository

    private var _binding:  FragmentPowerBinding? = null
    // This property is only valid between onCreateView and
// onDestroyView.
    private val binding get() = _binding!!

    private val model: RoutingServiceViewModel by viewModels()


    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentPowerBinding.inflate(layoutInflater)
        binding.toggleButton.setOnCheckedChangeListener { compoundButton: CompoundButton, b: Boolean ->
            lifecycleScope.softCancelLaunch {
                if (isActive) {
                    try {
                        if (b) {
                            serviceConnectionRepository.startService()
                            serviceConnectionRepository.bindService()
                        } else {
                            serviceConnectionRepository.stopService()
                            serviceConnectionRepository.unbindService()
                        }
                    } catch (e: IllegalStateException) {
                        compoundButton.isChecked = false
                        e.printStackTrace()
                    }
                }
            }
        }
        model.serviceConnections
                .observe(viewLifecycleOwner) {b -> binding.toggleButton.isChecked = b}
        return binding.root
    }

    companion object {
        private const val TAG = "PowerFragment"
    }
}